crates/ra_hir_ty/src/expr.rs

```diff
@@ -310,3 +315,112 @@ pub fn record_pattern_missing_fields(
     }
     Some((variant_def, missed_fields, exhaustive))
 }
+
+pub struct UnsafeExpr {
+    expr: ExprId,
+    inside_unsafe_block: bool,
+}
+
+impl UnsafeExpr {
+    fn new(expr: ExprId) -> Self {
+        Self { expr, inside_unsafe_block: false }
+    }
+}
+
+pub fn unsafe_expressions(
+    db: &dyn HirDatabase,
+    infer: &InferenceResult,
+    def: DefWithBodyId,
+) -> Vec<UnsafeExpr> {
+    let mut unsafe_exprs = vec![];
+    let mut unsafe_block_scopes = vec![];
+    let body = db.body(def);
+    let expr_scopes = db.expr_scopes(def);
+    for (id, expr) in body.exprs.iter() {
+        match expr {
+            Expr::Unsafe { body } => {
+                if let Some(scope) = expr_scopes.scope_for(*body) {
+                    unsafe_block_scopes.push(scope);
+                }
+            }
+            Expr::Call { callee, .. } => {
+                let ty = &infer.type_of_expr[*callee];
+                if let &Ty::Apply(ApplicationTy {
+                    ctor: TypeCtor::FnDef(CallableDef::FunctionId(func)),
+                    ..
+                }) = ty
+                {
+                    if db.function_data(func).is_unsafe {
+                        unsafe_exprs.push(UnsafeExpr::new(id));
+                    }
+                }
+            }
+            Expr::MethodCall { .. } => {
+                if infer
+                    .method_resolution(id)
+                    .map(|func| db.function_data(func).is_unsafe)
+                    .unwrap_or_else(|| false)
```

Comment:
`unwrap_or(false)` is simpler (and not slower)

Reply:
Woops, I think this was left around from when I was trying to do something unnecessarily complicated. I'll clean it up.